external/avro/src/test/scala/org/apache/spark/sql/avro/AvroEncoderSuite.scala

```diff
@@ -0,0 +1,352 @@
+        |    "default" : [ 1, 2, 3 ]
+        |  }, {
+        |    "name" : "string_array",
+        |    "type" : {
+        |      "type" : "array",
+        |      "items" : "string"
+        |    },
+        |    "default" : [ "a", "b", "c" ]
+        |  }, {
+        |    "name" : "record_array",
+        |    "type" : {
+        |      "type" : "array",
+        |      "items" : "simple_record"
+        |    },
+        |    "default" : [ {
+        |      "nested1" : 0,
+        |      "nested2" : "string"
+        |    }, {
+        |      "nested1" : 0,
+        |      "nested2" : "string"
+        |    } ]
+        |  }, {
+        |    "name" : "enum_array",
+        |    "type" : {
+        |      "type" : "array",
+        |      "items" : "simple_enums"
+        |    },
+        |    "default" : [ "SPADES", "HEARTS", "SPADES" ]
+        |  }, {
+        |    "name" : "fixed_array",
+        |    "type" : {
+        |      "type" : "array",
+        |      "items" : {
+        |        "type" : "fixed",
+        |        "name" : "simple_fixed",
+        |        "size" : 3
+        |      }
+        |    },
+        |    "default" : [ "foo", "bar", "baz" ]
+        |  }, {
+        |    "name" : "fixed",
+        |    "type" : {
+        |      "type" : "fixed",
+        |      "name" : "simple_fixed_item",
+        |      "size" : 16
+        |    },
+        |    "default" : "string_length_16"
+        |  }, {
+        |    "name" : "map",
+        |    "type" : {
+        |      "type" : "map",
+        |      "values" : "string"
+        |    },
+        |    "default" : {
+        |      "a" : "A"
+        |    }
+        |  } ]
+        |}
+      """.stripMargin
+
+    val schema = new Schema.Parser().parse(avroSchema)
+    val encoder = AvroEncoder.of[GenericData.Record](schema)
+    logInfo(schema.toString)
+    val expressionEncoder = encoder.asInstanceOf[ExpressionEncoder[GenericData.Record]]
+    val record = new GenericRecordBuilder(schema).build
+    val row = expressionEncoder.toRow(record)
+    val recordFromRow = expressionEncoder.resolveAndBind().fromRow(row)
+    assert(record.toString == recordFromRow.toString)
+  }
+
+  test("encoder resolves union types to rows") {
+    val avroSchema =
+      """
+        |{
+        |  "type" : "record",
+        |  "name" : "record",
+        |  "fields" : [ {
+        |    "name" : "int_null_union",
+        |    "type" : [ "null", "int" ],
+        |    "default" : null
+        |  }, {
+        |    "name" : "string_null_union",
+        |    "type" : [ "null", "string" ],
+        |    "default" : null
+        |  }, {
+        |    "name" : "int_long_union",
+        |    "type" : [ "int", "long" ],
+        |    "default" : 0
+        |  }, {
+        |    "name" : "float_double_union",
+        |    "type" : [ "float", "double" ],
+        |    "default" : 0.0
+        |  } ]
+        |}
+      """.stripMargin
+
+    val schema = new Schema.Parser().parse(avroSchema)
```

Comment:
nit: use the api using `jsonFormatSchema`

Reply:
Thanks, done in 00cb983.